src/HazelcastClient.ts

```diff
@@ -133,21 +133,86 @@ export class HazelcastClient {
         this.loggingService = new LoggingService(this.config.customLogger,
             this.config.properties['hazelcast.logging.level'] as string);
         this.loadBalancer = this.initLoadBalancer();
-        this.listenerService = new ListenerService(this);
+        this.errorFactory = new ClientErrorFactory();
+
+        this.listenerService = new ListenerService(
+            this.getLoggingService().getLogger(),
+            this.getConfig().network.smartRouting,
+            () => this.getConnectionManager(),
+            () => this.getInvocationService()
+        );
+
         this.serializationService = new SerializationServiceV1(this.config.serialization);
-        this.nearCacheManager = new NearCacheManager(this);
-        this.partitionService = new PartitionServiceImpl(this);
-        this.lifecycleService = new LifecycleServiceImpl(this);
+        this.nearCacheManager = new NearCacheManager(
+            this.getConfig(),
+            () => this.getSerializationService()
+        );
+
+        this.partitionService = new PartitionServiceImpl(
+            this.getLoggingService().getLogger(),
+            () => this.getSerializationService()
+        );
+        this.lifecycleService = new LifecycleServiceImpl(
+            this.getLoggingService().getLogger(),
+            this.getConfig().lifecycleListeners,
+            () => this.doShutdown(),
```

Comment:
As one of the simplest approaches to breaking cyclic dependencies, a special interface (or a type) could be introduced in `LifecycleService` (feel free to use different names for interface and method):
```ts
interface ShutdownableClient {

    doShutdown(): Promise<void>;
}
```

Then `LifecycleService` can expect an instance of `ShutdownableClient` in its constructor. This clearly describes the contract between the client and the service and allows for simplified mocking in tests.

The same approach can be used in other similar situations, like `ClientConnectionManager`.

Reply:
The main reason why I used closures is cyclic dependencies. I will try to do what you said

Reply:
Yes, cyclic dependencies is a problem for this particular refactoring. It would be great to get rid of cyclic dependencies where possible. This would help with making the structure of client's internal modules more granular and testable.

Reply:
Would not we pass the client itself to ListenerService in your proposition? What will be the benefit?

Reply:
The benefit is that `LifecycleService` will only depend (and consume) certain parts of client's API. As an alternative option here, we could think of moving the responsibility of shutting down the client into `HazelcastClient` class.